### PR TITLE
detect/asn1: Fix relative_offset keyword option - v1

### PR DESCRIPTION
**Pull request by sharksforarms**
- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [ ] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: https://redmine.openinfosecfoundation.org/issues/3720

Describe changes:
- Fix relative_offset keyword option to be relative in regards to the
last content match
- Change relative_offset to int32_t with bounds check to allow the full
range of the packet buffer size (uint16_t)
- Added checks for over/underflows
- Changed the offset type to uint16_t because the offset is applied to
the payload length, which is a uint16_t
- Adjusted test cases to work relative to the content match
- Added test case to verify bounds

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output (if applicable):

Note: I fixed this in https://github.com/OISF/suricata/pull/4954 as well

### CODE REVIEW

File: src/detect-asn1.h [hunk 1]
```diff
@@ -36,12 +36,11 @@
 typedef struct DetectAsn1Data_ {
     uint8_t flags;     /* flags indicating the checks loaded */
     uint32_t oversize_length;   /* Length argument if needed */
-    int32_t absolute_offset;   /* Length argument if needed */
+    uint16_t absolute_offset;   /* Length argument if needed */
```
Comment (jlucovsky) [inline src/detect-asn1.h:39]: What the rationale for changing the type?
Why leave the type of `relative_offset` unchanged?

File: src/detect-asn1.c [hunk 2]
```diff
@@ -148,15 +148,37 @@ static int DetectAsn1Match(DetectEngineThreadCtx *det_ctx, Packet *p,
...
-    int32_t offset;
+    uint16_t offset;
     if (ad->flags & ASN1_ABSOLUTE_OFFSET) {
         offset = ad->absolute_offset;
     } else if (ad->flags & ASN1_RELATIVE_OFFSET) {
-        offset = ad->relative_offset;
+        // relative offset in regards to the last content match
+
+        // Range check the relative_offset before downcasting to uint16_t
+        if (ad->relative_offset > UINT16_MAX
```
[the remark this answers is not in the document]
Comment (sharksforarms) [reply src/detect-asn1.c:158]: Range checking again as a defensive measure before downcasting, I can remove this check however as relative_offset is never touched otherwise

File: src/detect-asn1.h (hunk 1 above)
Comment (sharksforarms) [reply src/detect-asn1.h:39]: absolute_offset: Represents an offset from [0, payload_len], `Packet::payload_len` is represented as a `uint16_t`, so [0, uint16::max]
relative_offset: Represents an offset relative to the last content match which is between [0, uint16::max] so relative to this, we need a type to hold the range [-uint16::max, uint16::max] which is why it's int32, with a bounds check on parsing and also contains overflow checks

That was my train of thought, open to suggestions :)

File: src/detect-asn1.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (victorjulien) [reply src/detect-asn1.c:158]: If we're positive it's been checked at parsing we would use `BUG_ON()` instead.

File: src/detect-asn1.c [hunk 3]
```diff
@@ -148,15 +148,37 @@ static int DetectAsn1Match(DetectEngineThreadCtx *det_ctx, Packet *p,
...
+            || ad->relative_offset < -UINT16_MAX) {
+                return 0;
+        }
+
+        // Check for over/underflows
```
Comment (victorjulien) [inline src/detect-asn1.c:163]: I find these checks hard to follow. I think I would feel better seeing the checks done as u32 (or i32) and then only down casting after the values have been validated. Easier to follow for me anyway.

Comment (sharksforarms) [reply src/detect-asn1.c:163]: Agreed, re-reading this I get confused. I'll see if I can come up with a better solution

File: src/detect-asn1.h (hunk 1 above)
Comment (victorjulien) [reply src/detect-asn1.h:39]: Makes sense to me.

**Comment by sharksforarms** [comment]
https://github.com/OISF/suricata/pull/4978
